### ISSUE TITLE
Squash migrations

### ISSUE THREAD

**Issue by rowanmiller** [aspnet/EntityFramework#2174]
It would be good to have the ability to squash several migrations into a single file to help reduce the number of files in a project.

We probably want to keep track of the original list of migration names so that we can reason about this when targeting an existing database that the original migrations were applied to in their un-squashed form.

**Comment by popcatalin81** [comment]
@rowanmiller how will squash work? Merge Migrations into one or simply place them in a single file.

I think merging will be problematic when a target database has partial migrations but not all from a squash.

**Comment by bricelam** [comment]
@popcatalin81 I suspect at first, it will simpy concatenate all the operations together into one migration. In the future, it may try and simplify the operations (e.g. renaming A -> B -> C will become just A -> C)

Correct, "rewriting history" is always a bad idea. Before squashing, you'll have to unapply all the migrations you want to squash, squash them, then apply the new one. You shouldn't do it if the migrations have been applied on any database other than your local one.

This operation would be useful while developing a new feature. You could add all the migrations locally you want, but before merging you feature, you could squash them all down into a single migration.

**Comment by TheMadKow** [comment]
:+1:  For this idea

**Comment by markusvt** [comment]
I just wanted to suggest that idea also.. the migrations folder gets quite large quite fast if the projects develops over time

**Comment by dario-hd** [comment]
I was wondering if removing them all and creating an "initial migration" would be a better approach. In the end this is want I did recently. Of course the "initial migration" should be executed only on database creation. This will not only reduce the number of files in the project but it will also speed up the initial database creation if you recreate it multiple times e.g. for development and testing purposes.

**Comment by pgrm** [comment]
I'm puzzled that this issue is so inactive. How are others solving the issue of the ever growing Migrations folder?

Could there be at least some best practice described in the documentation @AndriySvyryd ?

**Comment by bricelam** [comment]
@pgrm I briefly mentioned a strategy in [the Migrations docs](https://github.com/bricelam/EntityFramework.Docs/blob/migrate/entity-framework/core/managing-schemas/migrations/projects.md) I'm adding...

**Comment by replaysMike** [comment]
This is a common problem we run into every once in a while. It's quite simple to accomplish. If your database is already up to date just delete all the migration files and truncate the dbo.__EFMigrationsHistory table. Generate a new initial create migration and you have now squashed all your migrations. You lose any comments but that's minor if you're needing to do it.

**Comment by PMExtra** [comment]
@replaysMike It will lost my custom migration operations. (For example, I set a custom default value for a new field.)

**Comment by replaysMike** [comment]
@PMExtra that’s surprising since you’re basically creating a migration based on the current state of the database. Is the default value being applied at the db level, or code level when the entity is created?

**Comment by deanvr** [comment]
Is this still open or closed? We often end up with way too many migration files in VS solution explorer, to resolve this its a pain in the... we manually run the migrations on a clean database, take the schema and data inserts and extract those and literally create a new migration that becomes the new seed, this way we don’t loose history and don’t have to truncate the migration table. What would be really cool is if we had a command that would reverse engineer a seed migration from a given database , this would save us a couple hours every month or so. Thank you - let’s make this happen!!!!!

**Comment by snebjorn** [comment]
I usually manually "squash" migrations into new InitialCreate when a new system goes into production, removing all the develop time migrations as they usually aren't very helpful to reason about model changes.

But changes to the model after the system is in production are very important.

I don't think the squashed migration should just be a concatenation of the existing migrations, as one might add a column/table and another might remove it again, it's just noise. It should be a fresh InitialCreate migration. Nice, clean and ready to run on the production environment without any develop time noise.

But this is already possible today, with possible data lose on the existing DEV environments.

I'd welcome a squash feature to make this easier and avoid data lose.

**Comment by roji** [comment]
@snebjorn if your goal is to squash development-time migrations which haven't yet been applied to production, it's pretty easy to simply reset your Migrations folder and then generate a single new one, which would contain all the changes - you can do that before merging your change.

**Comment by yakeer** [comment]
Any plans to automate this in any release?
The manual squashing is really frustrating.
And at some point, Azure DevOps is failing due to huge migrations dll.

**Comment by ajcvickers** [comment]
@yakeer This issue is open and on the backlog, which means that we do intend to implement this in a future release. See [release planning](https://docs.microsoft.com/en-us/ef/core/what-is-new/release-planning) for more details.

**Comment by clement911** [comment]
The latest VS release is extremely slow to compile migrations for example so this squashing might help.
See https://developercommunity.visualstudio.com/content/problem/1253845/visual-studio-2019-version-1680-very-slow-to-build.html

**Comment by roji** [comment]
@clement911 that issue should hopefully get resolved really soon on the VS side, so shouldn't require waiting for migration squashing in EF.

**Comment by RCTycooner** [comment]
@clement911  same issue here... going to attempt to manually squash my migrations now.

**Comment by clement911** [comment]
@RCTycooner this was fixed in the latest VS release 16.8.2

**Comment by RCTycooner** [comment]
@clement911 what's the status of issue for build pipelines? We're using hosted agents that should be getting the latest versions of everything, but still has this issue. Will update my VS to 16.8.2 and try a local build anyway

**Comment by domagojmedo** [comment]
Would squashing preserve any custom code that was added to the migration?

**Comment by MikaelEliasson** [comment]
I recently squashed a lot of migrations in our project (20+ person dev team). I hope this summary can help you or someone else with this need. See https://www.bokio.se/engineering-blog/how-to-squash-ef-core-migrations/ for more details + scripts we used.  

The largest hurdle I found doing this is that the migrations snapshots are actually "corrupt" in EF Core under certain circumstances. If you need me to dive more deeply into this just ping me.

Extract from the blog post:
```
This can happen when your team create migrations in parallel on different branches a basic flow is something like this.

Version Parent is on the dev branch.
Version B is branch from Parent and adds a migration.
Version C is branch from Parent and adds a migration.
Version D is B & C merged back into dev.

If we look at the global snapshot: 
B contains Parent + B
C contains Parent + C
D contains Parent + B + C

This is great. The problem is that no migration snapshot contains Parent + B + C. 

So if you pick B or C as you snapshot you will lose some information. Most of the time this is not a huge issue but EF uses these migrations snapshots to decide specificity of columns and if that needs to update. So I suddenly saw issues where on column had varchar(MAX) in the new version of the database but varchar(50) in the old one.
``` 

The issue this caused for me when squashing migrations was that I never could find a migration with a good migration snapshot to base it on. 

So instead I had to take one of two approaches:
- Use source control to pull the global snapshot from the time you want to  (not super easy to find)
- Manually merge the migration snapshot between the problematic ones.

If you find a solid solution for this issue squashing would actually be decently easy to script.

**Comment by Dasein732** [comment]
What I end up doing most of the time is delete the migrations and re add them to a fresh DB(almost identical to one of the articles mentioned in this thread, seeing how multiple people have a similar approach might be useful path to pursue) and then regenerate a single 'initial migration' script. One thing that would be extremely valuable in this situation would be to have a script that would do this for me(maybe have it optionally target an in memory provider for simplicity sake, since in 99% we're targeting a fresh DB instance) and insert a sort of a history breakpoint between different migrations in the Up()/Down() methods.

For example if we had 2 migrations before the squash, InitialMigration and AddCustomEntity, we'd have a migrationBuilder.InsertHistory("AddCustomEntity") in between those 2 migrations in the newly merged file and this would(optionally) trigger a transaction commit for previous migration.

This would be extremely useful, since the various deployments we cover usually have partial migrations applied due to various factors and we'll prolly never be able to bring them all to same verison.

**Comment by worthy7** [comment]
Just an idea but would this work?

1. Delete all migrations in the project, essentially reset
2. Create just the init migration
3. generate the idempotent SQL script
4. in the database, remove everything in the MigraitionsHistory table
5. Run just the last part of the SQL script, which should add the one and only migration info to the MigrationHistory table

If my understanding is correct, this should just work right?

**Comment by gojanpaolo** [comment]
@worthy7 That probably won't work as expected.

> https://github.com/dotnet/efcore/issues/2174#issuecomment-482414336
> It will lost my custom migration operations. (For example, I set a custom default value for a new field.)

**Comment by worthy7** [comment]
Ah, of course. I knew I was missing something.

**Comment by MikaelEliasson** [comment]
Just to leave a small update on my answer above. https://github.com/dotnet/efcore/issues/2174#issuecomment-760022208

We ran this for a second time. **And saved 1.8 million LOC (73MB of source code)**. This time it took about 60 min work to squash all the migrations, validate that the db was identical + add the test below.

We did run into a tiny issue though. The prep migration we created last time was included in the new initial migration. And it doesn't work in that context (duplicate key). We just commented it out.

```
![image](https://user-images.githubusercontent.com/888296/141791736-8d6a10b7-747b-495f-b3b1-bc2e209f48a3.png)

``` 

To prevent any team mate merging an earlier migration from a long lived branch we also added a small test that will block that in the PR tests.

```

        [Fact]
        public void NoMigrationShouldBeOlderThanSnapshot()
        {
            /**
             * When we squash changes we need to make sure no migrations before the squash is merged in.
             * 
             * If you have a migration like that you need to remove it on your branch and re-run the add-migration
             * so you get a newer timestamp. (And please merge newer changes into your branch first too) 
             */

            var currentActiveSnapshot = "20210826073610_Squash2_prep";

            TC.New();
            using (var context = DB.NewUnsafe())
            {
                var migrationsAssembly = context.GetService<IMigrationsAssembly>();

                var earlierMigrations = migrationsAssembly.Migrations.Where(x => x.Key.CompareTo(currentActiveSnapshot) < 0).Select(x => x.Key).ToList();
                earlierMigrations.ShouldBeEmpty("There are migrations that are older than latest squash.");
            }
        }
```

**Comment by voroninp** [comment]
![image](https://user-images.githubusercontent.com/9202028/181203106-15c9805e-a962-4f47-816b-4e69ae9dcb97.png)

**Comment by begerard** [comment]
Even a partial support witch would cover only the simple cases where there is no custom code required for the initial initialization would be useful. Like an automation of the workflow described by worthy7.
Giving us no automated solution even for simple cases make it always error prone, and make it by itself a defect in EF Core IMO.

**Comment by roji** [comment]
Just to be sure everyone is on the same page, especially for the simple cases with no custom migration code, there's a very easy workaround documented [here](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/managing?tabs=dotnet-core-cli#resetting-all-migrations); this is obviously not as perfect as a built-in squash feature, but the fact that a good workaround exists makes this less urgent.

The main problem IMHO is what happens when there **is** custom migration code, which gets lost if you simply delete the Migrations folder. In other words, I think there's very little value in implementing a "partial" squash feature which doesn't handle the custom code scenario.

In any case, this is feature which we definitely intend to implement at some point; it's simply a matter of competing priorities compared to other issues, which may not have a workaround as above. It's also pretty highly-voted, so there's a good chance we'll get to it sooner rather than later.

**Comment by worthy7** [comment]
I think tbh, this is probably something for the CLI. And in that case, it will just be a wrapper single function around the 6 steps I wrote above - with a warning that it doesn't care about custom code migrations 😂

**Comment by begerard** [comment]
> Just to be sure everyone is on the same page, especially for the simple cases with no custom migration code, there's a very easy workaround documented [here](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/managing?tabs=dotnet-core-cli#resetting-all-migrations); this is obviously not as perfect as a built-in squash feature, but the fact that a good workaround exists makes this less urgent.

The step where we delete/insert on the historical table ourself is not great. Time consuming, error prone (there is many way to do it)...
Having a simple command  "dotnet ef migrations reset" that handle the files removing/migrations generation (if that's a good way to do it) would be a very useful first step.

**Comment by MikaelEliasson** [comment]
Deleting all migrations is not a good workaround, it will be painful in any larger team/codebase to try that. Handling the merging of migrations coming from different branches will be really time consuming and error prone.

The approach I've written about is not perfect or super quick. But it handles both custom code migrations and a team environment. So there is workarounds even if none is perfect. 

If people go with the "simple" solution they should really artificially decrease the timestamp on the new migration a lot. That way you can at least avoid any merged migrations getting a lower timestamp and run before the initial migration as that will break things in a majority of cases. (And the documentation should probably include this warning too)

**Comment by roji** [comment]
One more thing to consider for this is mobile devices, where there's a (Sqlite) database on each device. Squashing is problematic, since databases are at various migration states, and you can only squash once you're none of your databases is in a migration state that's going to get squashed.

So squashing may be irrelevant for these scenarios. Another possible solution is to have a squash "archive" or some sort, which is somehow used only when a migration state in the database isn't found in source code. That would allow the main project to squash and be clean, but still allowing devices on old migration states to move forward. But this isn't trivial.

**Comment by stap123** [comment]
@roji Is that a common scenario for people? 

If the "normal" scenario would benefit from this as it seems it would is it a suitable option to add something and just advise that it's not suitable for all workloads/scenarios? (not sure what the teams general feeling on things like that is) 😄

**Comment by roji** [comment]
@stap123 there certainly are usages of EF on mobile devices out there. The point of my comment was mainly to keep these in mind when designing for this feature - we may end up saying it's not supported/recommended for mobile devices at first, possibly provide a solution later, etc.

**Comment by jeff-pang** [comment]
Other than snapshot issue explained above by @MikaelEliasson 

In order to preserve snapshots by using `migration remove` to remove migrations then squashing them runs the risk of data loss when renaming and dropping columns in the same table. Example:

1. Initial Migration - migration add InitialCreate` (applied to database)
```c# 
public class Customer
{
    public string Id { get; set; } = null!;
    public string? FirstName { get; set; }
    public string? LastName { get; set; }
}
```

2. Drop LastName - `migrations add DropLastName` (pending)
3. Rename FirstName to FullName - `migrations add RenameFirstName` (pending)

At this point we have `InitialCreate` applied to database and `DropLastName`, `RenameFirstName` pending which will be correct when applied to database.

But if we squash `DropLastName` and `RenameFirstName` by `migrations remove` and then recreating the migrations (e.g `migration add `RefactorCustomer`) we end up with migration which will drop and rename the wrong columns

```c#
 protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.DropColumn(
                name: "FirstName",
                schema: "sample",
                table: "Customers");

            migrationBuilder.RenameColumn(
                name: "LastName",
                schema: "sample",
                table: "Customers",
                newName: "FullName");
        }
```

**Comment by jasekiw** [comment]
@ajcvickers I believe more priority needs to be added to this task. Many of us have a very large amount of migrations that are causing our build times to be quite unreasonable slowing down our development speed. Simply deleting all migrations and recreating breaks several things: column order, custom migration adjustments, and more as mentioned by others above. Can this be added to the table of discussion for 8.0 or even 9.0 and commit to it? It has been punted several times.

**Comment by aradalvand** [comment]
@ajcvickers @roji Hey guys, would you please consider this for 8.0? It's one of the oldest and also most upvoted feature requests in the entire repo.

**Comment by MikaelEliasson** [comment]
One possible idea that doesn't solve the problem for old code bases but might make it easier to handle in the future. 

What if we could make `Add-Migration` also write a list of the migrations that was there when you ran the command? It's a very small change and even something you can implement as you own script that wraps `Add-Migration`. The result could basically be a plain text file next to the migration file with all the migrations listed row by row.

So why would that help? It would make it a lot easier to rebuild the actual migration graph and to understand what we might be missing in the snapshot for `MigrationX`. 

If you know what's missing it seems quite managable to rebuild a correct first snapshot. 

The same files/data could also likely be used to avoid the fact that `Remove-Migration` is completely broken in a team environment and must never be used today (because of the out of sync migration snapshots i creates a corrupted main snapshot 90% of the time). With that data it would likely be possible to figure if some migration is missing and maybe even how to handle it.

**Comment by roji** [comment]
@MikaelEliasson there's already `dotnet ef migrations list` for listing migrations. Note that there's no migration "graph" - it's just a list.

**Comment by inlineHamed** [comment]
This is my routine to keep build time short: Once a while I create an empty database and run the migrations to a desired point, then use SSMS to generate sql file (schema and data) from the database, and put it in migration folder, then move those migrations file into a zip file in migration folder.
Any new team member can just run the sql file instead of the migrations in zip folder.

**Comment by ajcvickers** [comment]
@jasekiw @aradalvand This is high up on our priority list, but unfortunately it is not trivial and requires expertise in the Migrations code. There are also other areas that for internal and strategic reasons need resources at the current time--see [Release planning process](https://learn.microsoft.com/en-us/ef/core/what-is-new/release-planning) for a general discussion of this. This means that its unfortunately not something we can work on in the 8.0 timeframe.

**Comment by dragnilar** [comment]
@ajcvickers  That's unfortunate, but understandable. I know it's too early to make any promises, but here's hoping we can see it for 9.0. 🥺🙏😇

**Comment by MoishyS** [comment]
I think most of the bloat is from the Designer files, and I don't understand why it's needed. 

is it only for removing migrations to recreate the main snapshot? can we safely remove those Designer files if we're not using the `migrations remove`? 
can we have a flag to not create them?

**Comment by ajcvickers** [comment]
@MoishyS I discussed this with the team. The designer files are frequently needed when executing migrations to obtain information from the underlying EF model. While in some cases it could be safe to remove them, this won't be the common case. It's also the case that a designer file not used by a migration in a given version of EF may later make use of the designer file as new features are implemented and bugs are fixed. Therefore, it doesn't seem like removing the designer files is a good way to go here.

**Comment by ajcvickers** [comment]
@AndriySvyryd @bricelam Do we have an existing issue for using the relational model here? I couldn't find one.

**Comment by AndriySvyryd** [comment]
@ajcvickers https://github.com/dotnet/efcore/issues/18620 😄

**Comment by roji** [comment]
That point in a project's life where everything has already been discussed...

**Comment by MoishyS** [comment]
@ajcvickers Thanks for your reply.
>The designer files are frequently needed when executing migrations to obtain information from the underlying EF model. 

I am trying to understand the case where you would need the snapshot for the migration, coming from Laravel all you have is the migration without any snapshot.
this behind-the-scene magic seems wrong to me, the migration file should include everything to create the sql without relying on anything else.

I understand that you need the snapshot history to remove migrations, but if someone does not use `migrations remove` (as we have everything tracked in git, and can reverse the main snapshot) we shouldn't need these huge designer files, and we shouldn't rely on it when executing migrations.

**Comment by ajcvickers** [comment]
@MoishyS It is not unreasonable to argue that everything needed should be in the migration, but this is currently not always the case. I don't have the specific cases immediately at hand, and they can also change from release to release for the same migration.

**Comment by bricelam** [comment]
One example is for table rebuilds on SQLite. The migration just contains operations like AlterColumn, AddForeignKey, AddCheckConstraint, etc. but SQLite doesn't support those operations natively, so we have to use the backing EF model to re-create the entire table.

**Comment by achikhv** [comment]
Also came to the point where we need to squash migrations. Build time and memory consumption became unacceptable.

We use idempotent migrations to update our customer's databases, and now my is:

1. Create full idempotent script that creates latest up-to-date database.
2. Delete all migrations and create new one.
3. Delete all content of the new migration and execute idempotent script in it.

It seems that this steps does the trick, and we can generate new idempotent script that is compatible with all versions of databases.

I would like to ask, are there any drawbacks in this plan? Will it be compatible with squash architecture that is in mind of EF Core team?

**Comment by jeff-pang** [comment]
> Also came to the point where we need to squash migrations. Build time and memory consumption became unacceptable.
> 
> We use idempotent migrations to update our customer's databases, and now my is:
> 
>     1. Create full idempotent script that creates latest up-to-date database.
> 
>     2. Delete all migrations and create new one.
> 
>     3. Delete all content of the new migration and execute idempotent script in it.
> 
> 
> It seems that this steps does the trick, and we can generate new idempotent script that is compatible with all versions of databases.
> 
> I would like to ask, are there any drawbacks in this plan? Will it be compatible with squash architecture that is in mind of EF Core team?

This comment perfectly captures how important to get this asap

1. Many of us are dealing with large migrations especially in a rapid development environment
2. Even if we think of a possible workaround we have no way to know if our workaround is safe

**Comment by achikhv** [comment]
> Also came to the point where we need to squash migrations. Build time and memory consumption became unacceptable.
> 
> We use idempotent migrations to update our customer's databases, and now my plan is:
> 
> 1. Create full idempotent script that creates latest up-to-date database.
> 2. Delete all migrations and create new one.
> 3. Delete all content of the new migration and execute idempotent script in it.
> 
> It seems that this steps does the trick, and we can generate new idempotent script that is compatible with all versions of databases.
> 
> I would like to ask, are there any drawbacks in this plan? Will it be compatible with squash architecture that is in mind of EF Core team?

To answer myself there is major drawback in postgres idempotent script. After all migrations squashed we need to create new idempotent script. SQL Server's script is good, but postgres created nested `DO $EF$ ... END $EF$; statements that caused syntax error when executed. So nesting idempotent scripts is not a great idea right now.

**Comment by roji** [comment]
@achikhv if you're encountering a syntax error with idempotent migrations on PostgreSQL, please open an issue on the PG provider repo with a minimal, runnable code sample.

**Comment by TsengSR** [comment]
@jeff-pang 
> This comment perfectly captures how important to get this asap
> 
> 1. Many of us are dealing with large migrations especially in a rapid development environment
> 2. Even if we think of a possible workaround we have no way to know if our workaround is safe

Well there is a viable workaround, which can be applied easily between major version releases. 

Let's say, for version 1.x you have your regular migrations and for 2.x you decide you want to "squash" it. 
1. When you first start tagging 2.x you run `dotnet ef database scripts` to export all of your migrations into a single SQL file.  
2. Optionally you also add an empty migration before that naming it `Initial_2_x`, more on later. Remember the name of your old Initial migration.
3. Then you drop all your migrations and create a new one from scratch (`dotnet ef migrations add Initial_2_x`)
4. If you did a empty `Initial_2_x` migration in 1.x release, then make sure your new `Initial_2_x` is named exactly the same as your previous version (including the leading timestamp). 

    The reason you may want to do that, is so that for **new** installations of your software the new Initial_2_x migration is used but for old upgrades it's skipped (EF Core compares the name of your migration if its in the `__EFMigrationHistory` table and if its there it will skip the migration)
5. Now, when your application starts, you need to check if the `Initial_1_x` exists, if so apply your exported migration script from step 1, because its an old installation and you need to bring the database up to the latest schema, before 2.x starts.

    If `Initial_1_x` isn't in the migration table, skip it (its a new installation) and EF Core will create the new schema with `Initial_2_x` instead. 

Works best between major releases. Keep the 1.x migration till 3.x and drop it, or replace it with a new one which includes 1.x and 2.x migration steps, if you want support migrations for more than 1 major upgrade.

This gives you the advantage, that even if a customer skips migrations, you can still upgrade the schema to the latest version before the "squash". And since the previous migrations are outsourced as sql files, they also do not need to be compiled anymore, just ship it with your release and apply it on the applications first start if required (upgrade installation only).

**Comment by TsengSR** [comment]
Because, let's face it. The possibility to downgrade is only used in two cases
1. Local development, to roll back a change that wasn't good or to squash it before the final commit. 
2. Once rolled out to production or the customer, reverting only happens by an compensation migration (remove field or table X)
3. When a customer wants to downgrade to an older version, because they are having an issue with the current one that can't be fixed quickly enough or can't afford to wait for an fix

First case is a non-case, as it never makes it to the main branch. 
Last one, may happen rarely, but shouldn't be an issue if the snapshots are "erased" with every major or minor release, since you'd rolll back the migration scripts back to the version you want to use and the version itself already has all the snappshots (removed earlier in the past when you squashed it)

**Comment by MikaelEliasson** [comment]
At Bokio we went for a mitigation to this issue where we placed our migrations in a separate project. But unlike the official documentation for this we were able to completely remove the reference to the migrations project from our app. We instead have a dedicated entry point project for running the migrations.

**The benefits of this approach:**
1) In dev we can unload the whole migrations project in VS. Which saves both a lot of memory, compile time and makes the searches in VS work better. 
2) The migrations are not included in our production builds meaning a smaller artifact and less memory consumption when the app is running.

1) is achievable with the approach recommended here https://learn.microsoft.com/en-us/ef/core/managing-schemas/migrations/projects?tabs=dotnet-core-cli
   
But to get the benefit of 2) we use a structure like this. I'll see if I can blog this later. But if you look at the dependencies you can see that the only project that refers to the migrations is a small helper project we use for adding and applying migrations and also generating migration scripts.   

![image](https://user-images.githubusercontent.com/888296/227721936-533f8be0-7cb5-464c-bb52-9c55692675c3.png)

**Comment by ErikEJ** [comment]
@MikaelEliasson This is great. You could even use a .slnf to filter out the migrations project.

**Comment by titobf** [comment]
Hi @MikaelEliasson. That's nice. How are then the migrations applied? (Who runs the MigrationHelper console app?) Thanks for sharing your solution.

**Comment by MikaelEliasson** [comment]
> Hi @MikaelEliasson. That's nice. How are then the migrations applied? (Who runs the MigrationHelper console app?) Thanks for sharing your solution.

In dev I as a developers apply them from the CLI just as usual. I have to change the command slightly though to point out the migrations project if I run them from `MigrationHelper` or the startup project if I run them from `Migrations`. We have some yarn shortcuts for doing things like this though so in day to day work I don't see the difference.

In our CI/CD pipelines it works like before. We just made the same change to the commands as above.

**Comment by ajcvickers** [comment]
Note for triage: we should consider documenting this approach: https://github.com/dotnet/efcore/issues/2174#issuecomment-1483833787

**Comment by begerard** [comment]
@MikaelEliasson That's an interesting approach, but if your Web project don't reference your Migrations project, you don't auto apply the migrations on a dev start (ie. dbContext.Database.Migrate();)? When another developer add a migration, he have to tell you to manually run the migrations?

**Comment by ErikEJ** [comment]
@begerard you should never use dbContext.Database.Migrate();)

**Comment by begerard** [comment]
> > @MikaelEliasson That's an interesting approach, but if your Web project don't reference your Migrations project, you don't auto apply the migrations on a dev start (ie. dbContext.Database.Migrate();)? When another developer add a migration, he have to tell you to manually run the migrations?
> 
> Technically you can do this as part of the build/deployment pipeline or as a [sidecar container](https://learn.microsoft.com/en-us/azure/architecture/patterns/sidecar) if you use docker/containers

I use this only in development ( if (app.Environment.IsDevelopment()) ), the CI do use the bundle (which is a sidecar indeed).

> @begerard you should never use dbContext.Database.Migrate();)

Isn't this the recommended practice for keeping a local dev database up to date? (I don't use docker locally)

**Comment by titobf** [comment]
> @begerard you should never use dbContext.Database.Migrate();)

Hi @ErikEJ. Can you briefly explain why it should never be used? Thanks

**Comment by roji** [comment]
@titobf more information on the shortcomings of Migrate() is available [in our docs](https://learn.microsoft.com/en-us/ef/core/managing-schemas/migrations/applying?tabs=dotnet-core-cli#apply-migrations-at-runtime).

**Comment by begerard** [comment]
> more information on the shortcomings of Migrate() is available [in our docs](https://learn.microsoft.com/en-us/ef/core/managing-schemas/migrations/applying?tabs=dotnet-core-cli#apply-migrations-at-runtime).

Thanks Roji, but if the .Migrate() is rightfully discouraged in production, is there better alternatives for local development? Maybe some docs helping configuring a dev env? I never found official resources on that.

**Comment by roji** [comment]
I'd guess the simplest way would be to simply use the CLI: `dotnet ef database update`.

**Comment by ajcvickers** [comment]
@begerard To add to what Shay said, I think it's reasonable to use `.Migrate()` in test environments, if that works for you. Usually, I find it easier to use `EnsureDeleted` and `EnsureCreated`. The point is that having the database automatically migration when the application starts has important drawbacks, especially in a multi-user environment where it happens when the first user makes a request.

**Comment by MichelZ** [comment]
I guess the `punted-for-6.0` label can be removed from this one

**Comment by roji** [comment]
@MichelZ we keep those on as a historical record for how many releases we punted on it, that can sometimes help us prioritize better.

**Comment by ronnyek** [comment]
I've used the delete the migrations folder and created a new migration that effectively bundles everything up into one... but I'm trying to have migrations between releases both up and down directions. 

Eg, I've got versioned releases 1.0.0, 1.0.1. I want to upgrade forward is easy, I can just run all migrations and my db will be upgraded, but if I then want to roll back all those changes that happened between 1.0.0 and 1.0.1, how do I do that?

I feel like that implies that as far as released code is concerned, migrations would be 1.0.0, 1.0.1 etc, but for development, it might be 1.0.0, <some feature implemented>, <some other feature implemented> but at the end of the day, I'd like to leave 1.0.0 in tact, and bundle those two migrations into a 1.0.1 migration.

I feel like migrations are generally very good, but I'm very confused about how people who bundle a version of their db with a service, and how they deal with a downgrade when a service is downgraded? Just deploy old code, and run migrate and it'll migrate to the older version of the db? (because the migrations assembly wont have those changes in the newer code?)

**Comment by pdevito3** [comment]
TIL this has been floating around for 7 years. I definitely could have missed it, but I’ve never seen it on any of the .NET roadmap plans even for consideration which is surprising given the number of upvotes. 

This is easily one of my biggest pain points I’ve just kind of lived with. It's obviously a hard problem, but I hope this makes it into one of the upcoming releases as it will be a huge value add and a great nod to promoting project longevity in the .NET ecosystem.

**Comment by yepeekai** [comment]
I read it all!!! In our case, as it was mention previously here, the designer.cs files are huge and there is one per migration which means A LOT.  We had a problem with azure build job where building the project containing the migrations would take an hour...  It has negative impact in visual studio on dev machines too.  I think those files should be representative of the associated .cs file: those should represent the increment instead of the complete model.  So a reference to the previous migration it is based on and then it should only include the changes.  Another solution (from a guy that doesn't know a lot about internal ef code) would be to have those designer.cs files compiled on demand instead of building them all ahead of time assuming most of the time, they are not needed, it would speed up most use cases.

**Comment by yepeekai** [comment]
out of the 8 most popular issues, this one is in third position and it is the only one that is not "consider for current release"

**Comment by jasekiw** [comment]
@ajcvickers I would like to ping you since I know that the planning for ef 9 has probably started. I know this is a really difficult task. Let us know if there is anything we can do to help.

**Comment by ajcvickers** [comment]
@jasekiw Unfortunately, the person who wrote all the Migrations code has now left the team. That isn't to say we won't still get to this at some point in the future, but I think it's unlikely to get done soon.

**Comment by yepeekai** [comment]
A part of me is sad... but I am, as always, grateful for this project transparency.  thanks.

**Comment by jasekiw** [comment]
@ajcvickers Thanks for the update! I really appreciate the transparency.

**Comment by voroninp** [comment]
@ajcvickers so you are a team of 4 now?

You, Shay, Andriy and Brice?

**Comment by ronnyek** [comment]
I feel like this is becoming the single most outstanding "missing or broken" features for me. The team has done an amazing job adding things we've needed to make efcore core component of our software.

Managing migrations over time has been really the only thing that has been problematic... or at least added lots of confusion as to how to deal with long term management of migrations over time best.

Is this something efcore team would be open to PR's on? Has any research or planning been done as far as the direction the team/framework intends to go with it?

**Comment by AndriySvyryd** [comment]
> You, Shay, Andriy and Brice?

Arthur, Shay, Maurycy and me

**Comment by titobf** [comment]
One good approach would be to have an "optional" switch to generate the migration directly on sql and without the designer file. For example, now I'm doing this manually. I add a migration (Add-Migration) and then use Script-Migration to generate the sql, drop it into the Up method (using migrationBuilder.Sql();) replacing the existing code, then copy the attributes DbContext and Migration from the designer file, and finally delete the designer file. This can be also done for the Down method. This way the migration becomes really lightweight.
The disadvantage is that you lock on the db engine and compatibility level, but it would be optional.

**Comment by TsengSR** [comment]
> The disadvantage is that you lock on the db engine and compatibility level, but it would be optional.

Migrations already lock you into a db and db engine, as of right now you already have to create multiple migrations for each database engine you intend to support, ideally in different assemblies

**Comment by ronnyek** [comment]
@ajcvickers Is there any opportunity where someone could potentially provide ideas/ submit PR's for this functionality? I wouldn't mind beginning to look into this... It's a massive problem we've run into time and time again, and wouldn't mind contributing some time and doing some research, if I thought it was something that might be evaluated for inclusion

**Comment by ajcvickers** [comment]
@ronnyek Yes, we're open to ideas. In particular, we are open radical ideas that don't involve this feature being implemented, since it is difficult to do right.  For example, the idea @titobf posted is interesting.

Things to consider:

- This could be something that deals with existing migrations, but this may be difficult because custom migrations are difficult to handle in an automated way.
- Instead, we could implement a new flavor of migrations which doesn't leave a lot of artifacts in its wake.
  - This system might have limitations. For example, it might be up-only.
  - This system should not have a backing model for each migration. This means that a migration cannot depend on the current model, which is relatively common today.

We were talking about this on the team last week, so it was interesting to see @titobf's idea.

The team is unlikely to have time to spend a lot of time in the near future, but we will pay attention to ideas posted here and give feedback. You may need to be patient; we are more busy than usual this year.

**Comment by ronnyek** [comment]
Forgive me for perhaps missing something... but validating my understanding here. (not talking custom migration steps here)
Make changes to context/entities
Create Migration (creates migration cs files for up/down)
Apply Migration (runs up(), and updates `__EFMigrationHistory` with the name of the latest applied migration)
![image](https://github.com/dotnet/efcore/assets/584713/43e424b4-aef4-459f-ac03-7df3fadaafd6)

My particular needs are basically collapsing many migrations into a release migration for a specific version of our software. While we'd potentially want to be able to roll back to the last version, we may not need granular rollback for everything that happened before last version was released.

Seems like you could effectively have a way of marking a particular migration as a snapshot, and then say when you want to publish next version of your app... some way of basically removing all the migrations back to last snapshot, and regenerating the migration from current model until now also marking the current migration as a snapshot.

Am I over simplifying? I realize removing migrations back to last snapshot and regenerating is probably overly simplistic, but I do think if you deleted records out of efmigrationhistory after previous snapshot (maybe without undoing the previous migrations... deleting columns/adding columns). 

I'm not sure how crazy custom migrations would need to work... but for me personally even if during the flatten, it just created the new migration with all the manual stuff that was happening before for seeding data or manual stuff, in the order it was happening... it could be a manual review type process.

**Comment by TomGathercole** [comment]
> @ronnyek Yes, we're open to ideas. In particular, we are open radical ideas that don't involve this feature being implemented, since it is difficult to do right. For example, the idea @titobf posted is interesting.
> 
> Things to consider:
> 
>     * This could be something that deals with existing migrations, but this may be difficult because custom migrations are difficult to handle in an automated way.
> 
>     * Instead, we could implement a new flavor of migrations which doesn't leave a lot of artifacts in its wake.
>       
>       * This system might have limitations. For example, it might be up-only.
>       * This system should not have a backing model for each migration. This means that a migration cannot depend on the current model, which is relatively common today.
> 
> 
> We were talking about this on the team last week, so it was interesting to see @titobf's idea.
> 
> The team is unlikely to have time to spend a lot of time in the near future, but we will pay attention to ideas posted here and give feedback. You may need to be patient; we are more busy than usual this year.

So we already have a process scripted to do this squash, but it feels super hacky. I've included the full script below, but the nastiest points are:
1. We have to time travel with git to generate a new snapshot from a particular point in time.
2. We have to do string manipulation to make a new migration look like the old `InitialMigration` so that the tooling doesn't try to re-apply it.
3. We have to do more string manipulation to add a trigger to the initial migration.

Now that last part is probably solved by [This Issue](https://github.com/dotnet/efcore/issues/10770), or more generally by not customizing anything in the migration that isn't strictly related to migrating _data_ from A to B. Also, I can't really imagine any general solution that could possibly solve that without just jamming all the migrations into one file.

For the other two points, I think these could potentially be solved by having some special handling around initial migrations. Honestly, I'm not entirely sure how this would work but potentially something like:
* Have some project level flag to opt into auto-generating an initial migration.
* This would always be generated with the same name (e.g. `00000000000000_Init`).
* If this migration is found in the history (or maybe if the __EFMigrationsHistory table exists full stop), then do nothing.
* Otherwise:
  * Find the first migration sequentially
  * Generate a 'default' migration based on the model snapshot of this migration

I don't actually think that last part would work, because the snapshot is the state _after_ that migration is applied, but perhaps someone smarter than me can figure out some way to make it work? I've only just realised it doesn't work, otherwise I maybe wouldn't have bothered typing those thoughts up...

Anyhow, that feels like it might be nice because it lets us randomly delete some series of early migrations, and things will 'just work', but you guys don't have to offer a guarantee that customized scripts will be handled in any way.


## Messy migration squash script, to illustrate the hoops we're jumping through!
Note, this is less opinionated version of our actual script, so this might not work properly. Hopefully it illustrates the broad strokes anyway.
```ps

Function First-Migration {
    Get-ChildItem $migrationFolder | Sort Basename | Select -First 1 -ExpandProperty Basename
}

Function Get-Class {    
    Param ($migration)
    $migration -replace '^[^_]*', ''
}

Function Rewrite-Migration {
    Param ($suffix)
    
	$source = "$migrationFolder\$newMigration$suffix"
	 
    $content = (Get-Content -Raw $source) `
        -replace $newMigration,$initialMigration `
        -replace (Get-Class $newMigration),(Get-Class $initialMigration) `
        -replace "(\s+}\s+protected override void Down)","$footer`$1" `
		
	Remove-Item $source
	
	return $content
}

Function Save-Migration {
    Param ($content, $suffix)

	$target = "$migrationFolder\$initialMigration$suffix"
    $content | Set-Content $target
}

Function Remove-IfExists {
	Param ($filesToRemove)

	ForEach ($f in $filesToRemove) {
		If (Test-Path $f) {
			Remove-Item $f
		}
	}
}

# 4 week sprints means that this covers two 28 day sprints.
# Technically we should only need one sprint, but this gives some contingency.
$days = 60

# Triggers cannot be managed in EF core yet, so we need to add these to the migration 'manually'
$footer = @" 
	migrationBuilder.Sql("CREATE OR ALTER TRIGGER MyTrigger ...");
"@

$migrationFolder = ".\MyProject\Migrations"

# Create a new branch for the job, updated to the latest main
git checkout main
git pull
git checkout -B "$jobNumber-squash-migrations"

# Get the name of the first migration file
$initialMigration = First-Migration

# Reset branch to the first migration created more than $days ago
$hash = git log --before="$days days ago" -1 --pretty=format:"%h" $migrationFolder
git reset --hard $hash

# Make a note of all the migrations which existed at this point, then delete them
# We want to delete the snapshot now, but ensure that we don't delete it a second time later on
$migrationsPattern = "$migrationFolder\*.cs"
$migrationsToDelete = Get-ChildItem $migrationsPattern -Exclude "*ContextModelSnapshot.cs"
Remove-Item $migrationsPattern

# Create a new initial migration
dotnet ef migrations add "$($jobNumber)_Squash"

# Get the name of the newly added migration
$newMigration = First-Migration

# Read out the new migration, make any replacements, then store in a varable
$rewrittenMigration = Rewrite-Migration ".cs"
$rewrittenDesigner = Rewrite-Migration ".Designer.cs"

# Reset back to main, and reapply the changes
git reset --hard main

Remove-IfExists $migrationsToDelete
Save-Migration $rewrittenMigration ".cs"
Save-Migration $rewrittenDesigner ".Designer.cs"

# Commit changes
git add .
git commit -m "Ref #$($jobNumber): Squash $stack migrations"
```

**Comment by ronnyek** [comment]
@TomGathercole I'm somewhat curious as to what role the git stuff here is providing. Is this simply as a mechanism to rollback migrations to an earlier state? Is this just some automation you are doing to auto collapse migrations every <time interval> ?

It seems to me like if you have a sort of order all the migrations were executed (whether that was by .cs file name) you could provide the last legitimate migration step, and collapse everything between that one.

Also seems like application of the migration may be weird since the db would have record of migrations that have been applied, but code/efcore cli tools wouldn't have anyway to tell what those migrations were.

**Comment by TomGathercole** [comment]
@ronnyek The purpose of the git stuff is to get to a point in history where the model generated from code matches some given 'initial migration' state. The main criteria is that all migration which make up this must have been applied to all environments which will ever be migrated going forward. 

There's definitely a fair bit of weirdness here. So far we haven't had any problems with it, since the__EFMigrationsHistory doesn't contain muchh aside from the names of the migrations. It would be nice to have some official-ish solution so that what we're doing (or something that achieves the same benefits) could be truly safe to rely on.

I think you might be on to something in regards to collapsing migrations. We could take the model in the .Designer.cs file for some last-known-applied-everywhere migration and re-generate the code based migration using this model. I don't think there's anything in the CLI tooling to do this, but there's probably something in the underlying library to generate part of a `ScaffoldedMigration` from an `IRelationalModel`

**Comment by deanvr** [comment]
Y'all are over thinking this... the problem statement is that on a real world project that spans years the migrations folder can get rather large and us poor dev folks end up manually cleaning up the folder by moving files to and archive folder. Any new dev often needs to get a DB backup and start from there... We need to be able to regenerate an initial migration from which a dev can use this to build up a full schema.

We need a command e.g. Add-Initial-Migration which runs the migrations and intentionally runs the model compare against an 'empty' schema, it then generates a migration class which could be run to generate the full schema upto that point. If the __EFMigrationsHistory table already contains migrations it will never run on update-database command, if for example you are a new dev and you run update database command and there are no existing __EFMigrationsHistory records or table it can run and generate the schema in the database from that migration class. 

Sounds easy peasy... ;)

**Comment by TomGathercole** [comment]
> Y'all are over thinking this... the problem statement is that on a real world project that spans years the migrations folder can get rather large and us poor dev folks end up manually cleaning up the folder by moving files to an archive folder. Any new dev often needs to get a DB backup and start from there... We need to be able to generate an initial migration from which a dev can use this to build up a full schema based off the current model.
> 
> We need a command e.g. Add-Initial-Migration which runs the migrations and intentionally runs the model compare against an 'empty' schema, it then generates a migration class which could be run to generate the full schema upto that point. If the __EFMigrationsHistory table already contains migrations it will never run on update-database command, if for example you are a new dev and you run update database command and there are no existing __EFMigrationsHistory records or table it can run and generate the schema in the database from that migration class.
> 
> Sounds easy peasy... ;)
> 
> Other option is some poor soul goes through all the migrations and literally combine all the ups, and all the downs into one migration :) Ohh couldn't we just get a code generator to do this mmmmm

This is essentially what we're doing, but the trouble is that there's no safe time to just delete all migrations and start again. At any given time, we'll usually have ~10 migrations pending deployment in a 2-weekly release which we cannot squash until they're deployed to every production environment.

I would absolutely be in favour of any solution that totally drops any manual migration tweaks though - I think that's going to be far too hard to deal with in a way that provides any real benefits.

**Comment by douglasg14b** [comment]
This would be helpful when doing development work.

I may be developing against a clone of staging, and making changes that generate migrations. I may make these changes incrementally, and running the application necessitates that I apply a migration for the changes made. The result is that while in development I may generate multiple migrations, and need to squash them all down to 1.

This is a tedious, error prone, process when done manually. An automated way to squash multiple migrations into 1 would be gold.

**Comment by diegopalat** [comment]
Any update on this topic?

**Comment by roji** [comment]
This issue is in the Backlog milestone. This means that it is not planned for the next release (EF Core 8.0). We will [re-assess the backlog](https://docs.microsoft.com/ef/core/what-is-new/release-planning) following the this release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources. Make sure to vote (👍) for this issue if it is important to you.

**Comment by pdevito3** [comment]
For anyone that's interested, I made a [.NET tool](https://github.com/pdevito3/StewardEF) that will squash (read, aggregate all your existing migrations) everything into your initial migration file. May still require some manual resolutions, but should still save a bunch of manual work.

**Comment by ErikEJ** [comment]
@pdevito3 you should get that added to the EF docs...

**Comment by jasekiw** [comment]
The tool does not work for me due to not handling ef core having a dependency on the target state of each migration. See the issue I created here: https://github.com/pdevito3/StewardEF/issues/1

I'm not sure if this is a database specific issues or not. I would assume not. I recommended another approach in the issue.

**Comment by diegopalat** [comment]
> For those who don't run migrations as part of app startup, you could just exclude /Migrations on CI (or even local) builds. For our particular setup this worked out quite well.

using which approach?

**Comment by jasekiw** [comment]
> For those who don't run migrations as part of app startup, you could just exclude /Migrations on CI (or even local) builds. For our particular setup this worked out quite well.

This is true but when interacting with migrations the build still slows down significantly. This also slows down packaging the migrations to be deployed. That's why we are still looking into ways to squash migrations.

**Comment by achikhv** [comment]
Maybe it should be more attention to https://github.com/dotnet/efcore/issues/18620? The core issue is that model snapshots take a lot of time and resources to compile. So I suppose that most of participants would come to terms with the number of migrations provided that migrations consume reasonable CPU and RAM to compile.

**Comment by aaref-sh** [comment]
I'd just created a new tool for that, called "EfMigrationsSquash"
explore it in this [repo](https://github.com/aaref-sh/EfMigrationsSquash)

> **How to use it**:
> - Download it using dotnet tool install --global EfMigrationsSquash if you don't have it already
> - cd to your migration project (or a project where do your migrations live)
> - run in terminal (EfMigrationsSquash), chose the start and end migration files to squash between
> - check the result

**Comment by jasekiw** [comment]
@aaref-sh Isn't this the same thing as the other project proposal mentioned a few comments up? https://github.com/dotnet/efcore/issues/2174#issuecomment-2469136795

**Comment by aaref-sh** [comment]
> [@aaref-sh](https://github.com/aaref-sh) Isn't this the same thing as the other project proposal mentioned a few comments up? [#2174 (comment)](https://github.com/dotnet/efcore/issues/2174#issuecomment-2469136795)

It looks yes, its similar I guess, however, I wasn't lucky enough to find this repo before you mention it, although I searched too long, 
but guess what! Hurray!!, now we have 2 tools to squash our migrations, isn't that awesome?

**Comment by yepeekai** [comment]
With Ai, I asked to put in place a solution to dynamically compile those designer.cs files on demand (roslyn).  In my case, I keep the last 5 ones intact and old ones are copied to an embedded text file per migration in the project.  It works fine, it adds a delay of about 5s per migration for 20000 lines designer files that needs to be compiled on demand.  It works for updating the database at the startup of the app or update-database or script-migration.  I am not going to share the code here, but it is easily reproducible with chat gpt or others.

**Comment by TsengSR** [comment]
@yepeekai 100% certain you didn't test it, because migrations are more complicated than just tossing everything into the same file. 

Up/Downs must work as before, also if you have your app at different versions at different customers, they all need be upgraded first to a specific version, otherwise you'd have errors.

i.e. if you have Migrations A, B and C at customer and you squash A, B, C, D into one file as `D'`, then deploy it, it will run on error because when you roll your squadshed migration to the customer he doesn't have D, but you have D squashed, so this migration will be missing from the customers system. 

What if you need to roll back a deployment? You can't roll to a specific version after you have squashed because your "down" method would roll back everything. Just some AI slop won't help you here.

**Comment by yepeekai** [comment]
I did not squash anything, all migrations are still there and each ones can execute on demand.  I did not try the down method since I don't use that, but mainly it hooks on the BuildTargetModel method in the designer.cs  Up definitely works.  I don't see why Down wouldn't.  I am talking about the designer.cs of each migration, which in our case is a problem since it increases CI compile time by a lot.  It is mainly to solve that without losing any functionality.

**Comment by mendozagit** [comment]
Roslyn is killing my machine 64GB RAM core i9 I have about 200 migrations anyone have been squashing migrations before?

**Comment by x-itm** [comment]
For the Roslyn/compile-time symptom, avoid rewriting migration history that may already exist in deployed databases. Squashing changes migration IDs and can break upgrades or targeted rollbacks unless every database is at the same known baseline.

A safer split is: keep the migration chain available for deployment/script generation, but avoid compiling it in builds that do not execute migrations (for example, a separate migration assembly or excluding it from the normal application build). If you do consolidate, treat it as a new baseline with an explicit cutover plan and validate `Script-Migration`/idempotent scripts plus rollback paths against representative databases.

**Comment by mendozagit** [comment]
The assembly/project/place doesn't matter, it needs to be compiled anyways, that's the problem.

**Comment by x-itm** [comment]
Agreed—the compiler still has to process every migration source file, so relocating the migrations only helps if it removes that project from the design-time/build graph you are invoking. If the migrations assembly is referenced and compiled on every build, it will not address Roslyn memory pressure.

For that case, the practical options are to reduce the generated source that must compile (for example, consolidate only a fully retired baseline after validating all deployed databases) or isolate migration tooling so normal application builds do not compile it. The former needs a safe cutover plan because existing databases retain the old migration IDs.
